### ISSUE TITLE
Replication vs ReplicasPerPartition

### ISSUE THREAD

**Issue by mayankshriv** [apache/pinot#8804]
The question on what is the difference between `replication` vs `replicasPerPartition` frequently comes up. Opening this issue to track:

- Could these be unified to a single config `replication` that applies to both `OFFLINE` as well as `REALTIME` tables? We can still honor `replicasPerPartition` for backward compatibility.
- If there is we need both, could we document the reasoning behind it?

**Comment by mcvsubbu** [comment]
Here is the history:
When we had only HLC, we had `replication` for both OFFLINE as well as REALTIME tables. In case of REALTIME, it was used to [derive](https://github.com/apache/pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotTableIdealStateBuilder.java#L89) the number of realtime servers to consume the topic and also whether the topic was divided amongst the servers (something that only Uber used, I believe -- to split events on same topic across servers).

Therefore, this config could not be used for LLC, and we picked a new config -- `replicasPerPartition`. Note that `replication` is still used for OFFLINE tables.

I am not sure if we can declare HLC dead as yet. We can certainly poll the community and check, and if everyone agrees, we can remove the code, and proceed to treat `replication` as the number of replicas -- whether offline or realtime.

Another way to do this would be to use `replication` as the number of replicas if a table has only LLC (but then we have to make sure that they never ever add HLC to it -- having both HLC and LLC is one way of migrating from HLC to LLC). Fall back to `replicasPerPartition` if that `replicas` is not defined. This may be a more palatable way in case people are using HLC.

Happy to discuss over a small doc rather than this issue. We can then float the doc to the public as well

**Comment by Jackie-Jiang** [comment]
@mcvsubbu Thanks for sharing the history. I think it is always safe for LLC table to first check `replicasPerPartition`, and if it is not defined, fall back to `replication`.

**Comment by mcvsubbu** [comment]
Are we declaring HLC dead? I am fine with that, except if some stream comes along that cannot support LLC (I think we had a bit of a hard time with gps right?)

**Comment by Jackie-Jiang** [comment]
> Are we declaring HLC dead? I am fine with that, except if some stream comes along that cannot support LLC (I think we had a bit of a hard time with gps right?)

No, we didn't yet, and that's why I was suggesting still first checking `replicasPerPartition` for LLC, and fall back to `replication` if it is not available.
Long term wise, we don't want to use HLC because there are lots of known issues around it, and it is not as reliable as LLC.

**Comment by mcvsubbu** [comment]
@yupeng9  can u please let us know if you are still using this setting?

**Comment by Jackie-Jiang** [comment]
cc @snleee

**Comment by yupeng9** [comment]
Missed this thread. Uber uses LLC only.
